common/predictive-text/build.sh

```diff
@@ -20,92 +20,147 @@ THIS_SCRIPT="$(greadlink -f "${BASH_SOURCE[0]}" 2>/dev/null || readlink -f "${BA
 # This script runs from its own folder
 cd "$(dirname "$THIS_SCRIPT")"
 
+################################ Main script ################################
+
+builder_check_color "$@"
+
+builder_describe "Builds the lm-layer module" \
+  "clean" \
+  "configure" \
+  "build" \
+  "test" \
+  ":libraries  Targets all in-repo libraries that this module is dependent upon" \
```

Comment:
Conceptually, `configure` is supposed to configure and build any internal and extenal dependencies, so then you don't need the `:libraries` target.

The primary issue here is: that `npm i`/`npm ci` is just so painfully slow and noisy, so we really can't afford to run it for each dependency, which would keep their `configure` action clean and isolated. If we can find a good clean method of only running `npm ci` _once_ per command-line invocation of `build.sh configure` (that is, if we called `verify_npm_setup` more than once, it would only run `npm ci` once), I think that really does cover most of our bases. This, I think, does it (and removes need for `fetch_deps` parameter to `verify_npm_setup`):

```bash
# Installs npm dependencies, but only once per build invocation
verify_npm_setup () {
  if [ -z ${_builder_verify_npm_setup_run+x} ]; then
    export _builder_verify_npm_setup_run=true

    # Check if Node.JS/npm is installed.
    type npm >/dev/null ||\
      fail "Build environment setup error detected!  Please ensure Node.js is installed!"

    init_npm
  fi
}
```

There is also overlap between `common/web/keyman-version/build.sh` and `tsc -b` locally, because `tsc -b` builds all dependencies if they are out of date. But as you know, keyman-version has a required pre-build step which `tsc -b` does not support. So we have to run the build script.

I am formulating a better approach which in theory would be passed into `builder_describe`. But have not yet completed that design...

Reply:
Part of the reason for the `:libraries` target is because of how it provides a wonderful dev-feature when unit testing the various web projects.  In particular, I personally want the `test:libraries` action to exist; isolating aspects of `test`-only build parts to a `build:libraries` is a happy add-on.

If I've been working on a bug in the main `/web/source` code, I could care less what test results I get from `keyboard_processor` and `input_processor` unit tests - they're completely irrelevant.  Much of the time, I'm also not looking to recompile any test-dependency code, as with Web's recorder modules/tools.  They're often "build once" as far as many parts of the codebase are concerned.

As far as `build:libraries` is concerned, one of the benefits is being able to bypass the `npm` step entirely.  Often, I know that there's absolutely no need for even a single `npm ci` call - and one call adds enough delay that I'd rather zero calls than even a single one.  So, with the current structure proposed by this PR's branch, I can skip the `configure` step in order to get zero such calls while still being able to rebuild in-repo parts of the codebase.  (Say, if I made changes to `lm-worker` code and am rebuilding the lm-layer (`common/predictive-text`.)

If the semantics worry you, perhaps something like the following might make sense?
- `configure-env` ("configure environment")
- `configure-libs` ("configure in-repo dependencies")

Though I suppose that would then necessitate an extra option if dropping `:libraries`:
- `--chain`, intended for use with `test`:  also runs in-repo dependency test scripts
    - The nomenclature can be workshopped, of course.

I feel like that proposed option is less elegant, but we could go that route if you prefer.

Reply:
So, I want targets to be things in the current script. Ideally, a given build.sh will figure out if internal or external dependencies need configuring or building, and only do those if necessary. It would generally never clean or test dependencies.

However, at present, I have not tried to solve the internal dependency question -- just relying on the various dependencies being configured/built as needed by specific temporary steps. And while what you've got here is a good step, I don't think it quite gets to where I want to be.

I will think about this in the next few days because we do need to cross this bridge for dependencies, and I'd like to get this right -- the rest of the builder scripts stuff is reasonably clean so want to try and keep it that way for internal dependencies also.

Reply:
I hear you.  I spent longer than I'd prefer to admit before landing at this PR's current "solution" as it was, trying to keep it "reasonably clean."  KMW will want to use a similar approach to whatever we land on here, and I could see similar concerns for the Android & iOS scripts.

Reply:
> similar concerns

... not to mention Developer which is arguably the worst of all, with dependencies all over the place from web, core, and common.

I'm too tired to be able to do a good job of this tonight, but will continue on this after my vacation.